### PR TITLE
[DSET-841] Build test and publish agent package to packagecloud

### PR DESCRIPTION
**Pull request by ArthurKamalov**
This pull request adds production versions of the agent package to the Packageloud.

Another big change is that current end-to-end (or so-called 'AMI' tests) have been fully rewritten in Python ([link](https://github.com/scalyr/scalyr-agent-2/blob/0896974089c106ac251e7f731f27c9d6463a7a98/tests/end_to_end_tests/managed_packages_tests/test_packages.py#L138)).

### CODE REVIEW

File: agent_build_refactored/managed_packages/convenience_install_script/install_agent_template.sh [hunk 1]
```diff
@@ -0,0 +1,432 @@
+#!/bin/bash
```
Comment (ArthurKamalov) [inline agent_build_refactored/managed_packages/convenience_install_script/install_agent_template.sh:1]: This is basically just a copy of [existing install script ](https://github.com/scalyr/scalyr-agent-2/blob/master/scripts/circleci/release/installScalyrAgentV2.sh) but without some legacy parts.

File: agent_build_refactored/managed_packages/files/init.d/scalyr-agent-2 [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Comment (ArthurKamalov) [inline agent_build_refactored/managed_packages/files/init.d/scalyr-agent-2:1]: Also Just a copy of existing file.

File: agent_build_refactored/managed_packages/install-scriptlets/preuninstall.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
```
Comment (ArthurKamalov) [inline agent_build_refactored/managed_packages/install-scriptlets/preuninstall.sh:1]: Full copy of the https://github.com/scalyr/scalyr-agent-2/blob/master/installer/scripts/preuninstall.sh

File: agent_build_refactored/scripts/filter_matrix_jobs.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
+# Copyright 2014-2022 Scalyr Inc.
```
Comment (ArthurKamalov) [inline agent_build_refactored/scripts/filter_matrix_jobs.py:1]: This basically just a part of the `agent_build_refactored/scripts/get_github_actions_job_matrices.py` file, which was logically split into three files.

File: agent_build_refactored/scripts/get_workflow_info.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
+# Copyright 2014-2022 Scalyr Inc.
```
Comment (ArthurKamalov) [inline agent_build_refactored/scripts/get_workflow_info.py:1]: This basically just a part of the `agent_build_refactored/scripts/get_github_actions_job_matrices.py` file, which was logically split into three files.

File: .github/actions/init-job-matrices/action.yml [hunk 6]
```diff
@@ -45,34 +55,38 @@ runs:
...
+        echo "to_publish=$(cat /tmp/workflow_info.json | jq -c '.to_publish')" >> $GITHUB_OUTPUT
+        echo "is_production=$(cat /tmp/workflow_info.json | jq -c '.is_production')" >> $GITHUB_OUTPUT
+        echo "version=$(cat /tmp/workflow_info.json | jq -c '.version')" >> $GITHUB_OUTPUT
+        
+        mkdir -p /tmp/matrices
+        
+        cat <<EOT | python3 agent_build_refactored/scripts/filter_matrix_jobs.py --is-master-run "${IS_MASTER_RUN}" > /tmp/matrices/agent_image_build_matrix.json
         ${{ inputs.images-to-build }}
         EOT
+        echo "agent_image_build_matrix_json=$(cat /tmp/matrices/agent_image_build_matrix.json)" >> $GITHUB_OUTPUT
```
Comment (jmakar-scalyr) [inline .github/actions/init-job-matrices/action.yml:76]: I think this is a litter clearer:
```
echo ${{ inputs.images-to-build }} | python3 agent_build_refactored/... > /tmp/matrices/...
echo "agent_image_build_matrix_json=$(cat /tmp/matrices/...)" >> $GITHUB_OUTPUT
```

If the temp files are not needed later could even do this:
```
echo "agent_image_build_matrix_json=$(echo ${{ inputs.image-to-build }} | python3 agent_build_refactored/...)" >> $GITHUB_OUTPUT
```
That may be less readable so I'll leave it up to you.

This overall comment applies to the other two variables below also.

File: .github/workflows/agent-build-refactored.yml [hunk 7]
```diff
@@ -23,9 +25,10 @@ jobs:
   pre_job:
     runs-on: ubuntu-20.04
     outputs:
-      should_skip: ${{ steps.skip_check.outputs.should_skip }}
```
Comment (jmakar-scalyr) [inline .github/workflows/agent-build-refactored.yml:26]: Out of curiosity, why was this removed / no longer necessary?

File: agent_build_refactored/docker_image_builders.py [hunk 8]
```diff
@@ -259,8 +259,8 @@ def build_container_filesystem(self, output_path: pl.Path):
             shutil.rmtree(package_root_path)
         package_root_path.mkdir(parents=True)
 
-        # Build 'LFS-structured' filesystem.
-        build_linux_lfs_agent_files(
+        # Build 'HFS-structured' filesystem.
```
Comment (jmakar-scalyr) [inline agent_build_refactored/docker_image_builders.py:262]: Looks like a typo here: HFS-structured => FHS-structured

By the way, FHS refers to https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard right?  What does LFS refer to?

File: agent_build_refactored/scripts/get_workflow_info.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+GITHUB_TOKEN = os.environ.get("GITHUB_TOKEN")
+GITHUB_SHA = os.environ.get("GITHUB_SHA", "")
+
+
+def is_branch_has_pull_requests() -> bool:
+    """
+    Determine if current branch has at least one opened pull request.
+    """
+    data = subprocess.check_output([
+        "curl",
```
Comment (jmakar-scalyr) [inline agent_build_refactored/scripts/get_workflow_info.py:50]: I think will also need a `-s` here to avoid showing the progress meter

File: agent_build_refactored/tools/runner.py [hunk 10]
```diff
@@ -831,7 +830,7 @@ def run_in_docker(
 
             mount_args.extend([
                 "-v",
-                f"{arg.path}:{in_docker_path}:z"
```
Comment (jmakar-scalyr) [inline agent_build_refactored/tools/runner.py:834]: Why is the :z no longer necessary?  It's no longer shared?

File: tests/end_to_end_tests/managed_packages_tests/test_packages.py [hunk 11]
```diff
@@ -167,81 +189,400 @@ def test_packages(
...
+    assert "ca_cert_path: /etc/ssl/system-ca-bundle.crt" in agent_log
+    assert "HttpConnection uses native os ssl" in agent_log
+
+    agent_status = agent_commander.get_status()
+    assert "Last successful communication with Scalyr:" in agent_status
+    assert "Last successful communication with Scalyr: Never" not in agent_status
+    assert "Last copy response status:                 success" in agent_status
+    _stop_agent_and_remove_logs_and_data(agent_commander=agent_commander)
+
+    # 3. Mimic MITM attack by pointing a random domain to agent.scalyr.com and verify hostname
```
Comment (jmakar-scalyr) [inline tests/end_to_end_tests/managed_packages_tests/test_packages.py:389]: Nicely done!

File: agent_build_refactored/scripts/filter_matrix_jobs.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+DEFAULT_OS = os.environ["DEFAULT_OS"]
+DEFAULT_PYTHON_VERSION = os.environ["DEFAULT_PYTHON_VERSION"]
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--is-master-run",
+        required=True
+    )
```
[the remark this answers is not in the document]
Comment (ArthurKamalov) [reply agent_build_refactored/scripts/filter_matrix_jobs.py:41]: Since we get input value for this option from JSON, then it can be `true` or `false`, in case of `false` the `store_true` option will still be true, so, on the "command caller" side, we must not provide the `--is-master-run`  at all, but that's also adds some extra verbosity.

I guess we can do that with bash's conditional substitution, (like `${value:--is-master-run}`) but in this case we still have to return empty string instead of `false` in our JSON object.

File: agent_build_refactored/tools/runner.py (hunk 10 above)
Comment (ArthurKamalov) [reply agent_build_refactored/tools/runner.py:834]: I, actually, haven't got any difference with or without it. I just added it in hope that it will solve user permission differences of host and guest OS's, but it didn't.

File: agent_build_refactored/docker_image_builders.py (hunk 8 above)
Comment (ArthurKamalov) [reply agent_build_refactored/docker_image_builders.py]: Oops, yes it does, fixed.  Thanks!

File: .github/workflows/agent-build-refactored.yml (hunk 7 above)
Comment (ArthurKamalov) [reply .github/workflows/agent-build-refactored.yml:26]: Because for this workflow I just left `opened`  event for the `pull_request` trigger type, https://github.com/scalyr/scalyr-agent-2/blob/0896974089c106ac251e7f731f27c9d6463a7a98/.github/workflows/agent-build-refactored.yml#L4-L7
 so there should not be any "double triggering" from both the `push` and `pull_request` when you push new commits to already opened pull request.

File: .github/actions/init-job-matrices/action.yml (hunk 6 above)
Comment (ArthurKamalov) [reply .github/actions/init-job-matrices/action.yml:76]: There reason is because that values are multiline string and with `echo ${{ inputs.images-to-build }}`, it just leads to bash syntactic error. And with `echo "${{ inputs.images-to-build }}"` it just removes all double quoted from the string, so it's not a valid JSON anymore.
But thanks for pointing on this again, because it seems that single quote does the trick `echo '${{ inputs.images-to-build }}'` 👍

File: tests/end_to_end_tests/managed_packages_tests/test_packages.py (hunk 11 above)
Comment (ArthurKamalov) [reply tests/end_to_end_tests/managed_packages_tests/test_packages.py:389]: Thanks, but that's already existing checks and I just moved them here :sweat_smile:
